frame/balances/src/lib.rs

```diff
@@ -900,7 +900,7 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		f: impl FnOnce(&mut AccountData<T::Balance>, bool) -> Result<R, E>,
 	) -> Result<(R, DustCleaner<T, I>), E> {
 		let result = T::AccountStore::try_mutate_exists(who, |maybe_account| {
-			let is_new = maybe_account.is_none();
+			let is_new = maybe_account.is_none() && !system::Pallet::<T>::account_exists(who);
```

Comment:
I didn't think about the whole picture and how this modification but this doesn't solve the whole picture I think.

at the end of your test below alice has 2 providers and 1 consumer.
Maybe it is ok when ED is 0 to consider that balance can increment providers indefinitely ?

If pallet balance has ED 0 then is there a way for an account to be deleted ? if there is no way to delete an account, then I guess having increments of provider doesn't harm.

Reply:
@liuchengxu do you think increment provider like this makes sense ?

Reply:
Hmm, it's fine with ED 0 from my point of view since the account will never be deleted anyway. But this PR specifically patches for these using system as the account store of balances, what if there is a project that uses another pallet instead of system? Even though the chances are pretty low, we'd better use a more general solution.

Reply:
@liuchengxu , system pallet as I know is a  base layer for all pallets from frame and any other new pallets which will be added into the runtime. Meanwhile in the pallet balances you can already see the usage of the system pallet so I do not think that it will be a problem.
https://github.com/paritytech/substrate/blob/4d28ebeb8b027ca0227fe7779c5beb70a7b56467/frame/balances/src/lib.rs#L827
https://github.com/paritytech/substrate/blob/4d28ebeb8b027ca0227fe7779c5beb70a7b56467/frame/balances/src/lib.rs#L956 etc.

Reply:
But I think increasing of the `providers` field it is not good and can lead to the problems.
Need deeper investigation of this problem.

Reply:
@Mr-Leshiy I meant the account store in balances pallet:

https://github.com/paritytech/substrate/blob/1b4b05d2c61b9da648a9124d9d52c82b012e6013/frame/balances/src/lib.rs#L216-L217

This patch does not work if some project stores the account in a non-system pallet. 

But I fully agree we should think more about this problem as it's pretty fundamental.

Reply:
My initial research have found the following.
According to this trait https://github.com/paritytech/substrate/blob/4d28ebeb8b027ca0227fe7779c5beb70a7b56467/frame/support/src/traits/stored_map.rs#L26 we can have `default` value of item at generally it does not considered as special case, because we can store a `default` item in the storage. But if we will remove something or reset it can be assigned to the `default` value.
Moving to the implementation of this trait, to this function https://github.com/paritytech/substrate/blob/4d28ebeb8b027ca0227fe7779c5beb70a7b56467/frame/system/src/lib.rs#L1654 , it seems we can see that here we are considering `default` value as special case that before we do not have any value that it is initialized the first time, that is why we have incremented `providers` field

Reply:
isn't it what is documented on the item:
```
/// Implement StoredMap for a simple single-item, provide-when-not-default system. This works fine
/// for storing a single item which allows the account to continue existing as long as it's not
/// empty/default.
///
/// Anything more complex will need more sophisticated logic.
impl<T: Config> StoredMap<T::AccountId, T::AccountData> for Pallet<T> {
```
The stored map is considered as a provide when not default.
Using an option here instead of relying on the default for currency would solve it. But requires migrating all the accounts which is annoying.